File: src/main/java/com/matag/game/cardinstance/ability/CardInstanceAbility.java
package com.matag.game.cardinstance.ability;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.matag.cards.Card;
import com.matag.cards.ability.Ability;
import com.matag.cards.ability.AbilityService;
import com.matag.cards.ability.AbilityTransposer;
import com.matag.cards.ability.selector.MagicInstanceSelector;
import com.matag.cards.ability.selector.SelectorType;
import com.matag.cards.ability.target.Target;
import com.matag.cards.ability.trigger.Trigger;
import com.matag.cards.ability.type.AbilityType;

import java.util.List;
import java.util.Optional;

import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toList;

public class CardInstanceAbility extends Ability {
  public CardInstanceAbility(Ability ability) {
    super(ability.getAbilityType(), ability.getTargets(), ability.getMagicInstanceSelector(), ability.getParameters(),
      ability.getTrigger(), ability.getAbility(), ability.isSorcerySpeed());
  }

  public CardInstanceAbility(AbilityType abilityType) {
    super(abilityType, emptyList(), null, emptyList(), null, null, false);
  }

  public CardInstanceAbility(AbilityType abilityType, List<Target> targets, List<String> parameters, Trigger trigger) {
    super(abilityType, targets, null, parameters, trigger, null, false);
  }

  public CardInstanceAbility(AbilityType abilityType, MagicInstanceSelector magicInstanceSelector, List<String> parameters, Trigger trigger) {
    super(abilityType, emptyList(), magicInstanceSelector, parameters, trigger, null, false);
  }

  @JsonProperty
  public String getAbilityTypeText() {
    String parametersString = new AbilityService().parametersAsString(parameters);

    if (abilityType == AbilityType.SELECTED_PERMANENTS_GET) {
      if (magicInstanceSelector.getSelectorType() == SelectorType.PLAYER) {
        return String.format(abilityType.getText(), magicInstanceSelector.getText(), parametersString) + ".";
      } else {
        return String.format(abilityType.getText(), magicInstanceSelector.getText(), parametersString) + " until end of turn.";
      }

    } else {
      return String.format(abilityType.getText(), parametersString);
    }
  }

  public CardInstanceAbility getAbility() {
    return getCardInstanceAbility(ability);
  }

  public String getParameter(int i) {
    if (parameters.size() > i) {
      return parameters.get(i);
    }
    return null;
  }

  public boolean requiresTarget() {
    return !targets.isEmpty();
  }

  public static List<CardInstanceAbility> getCardInstanceAbilities(Card card) {
    return Optional.ofNullable(card)
      .map(Card::getAbilities)
      .orElse(emptyList())
      .stream()
      .map(CardInstanceAbility::getCardInstanceAbility)
      .collect(toList());
  }

  public static CardInstanceAbility getCardInstanceAbility(Ability ability) {
    return Optional.ofNullable(ability)
      .map(AbilityTransposer::transpose)
      .map(CardInstanceAbility::new)
      .orElse(null);
  }
}

File: src/main/java/com/matag/game/cardinstance/ability/CardInstanceAbility.kt
package com.matag.game.cardinstance.ability

import com.fasterxml.jackson.annotation.JsonProperty
import com.matag.cards.Card
import com.matag.cards.ability.Ability
import com.matag.cards.ability.AbilityService
import com.matag.cards.ability.selector.SelectorType
import com.matag.cards.ability.transpose
import com.matag.cards.ability.type.AbilityType

data class CardInstanceAbility(
    val ability: Ability
) {
    constructor(abilityType: AbilityType) : this(Ability(abilityType))

    constructor(cardInstanceAbility: CardInstanceAbility) : this(
        Ability(
            cardInstanceAbility.ability.abilityType,
            cardInstanceAbility.ability.targets,
            cardInstanceAbility.ability.magicInstanceSelector,
            cardInstanceAbility.ability.parameters,
            cardInstanceAbility.ability.trigger,
            cardInstanceAbility.ability.ability,
            cardInstanceAbility.ability.sorcerySpeed,
            cardInstanceAbility.ability.optional
        )
    )

    @get:JsonProperty
    val abilityTypeText: String
        get() {
            val parametersString: String = AbilityService().parametersAsString(ability.parameters!!)

            if (ability.abilityType == AbilityType.SELECTED_PERMANENTS_GET) {
                val magicInstanceSelector = ability.magicInstanceSelector!!
                if (magicInstanceSelector.selectorType == SelectorType.PLAYER) {
                    return String.format(ability.abilityType.text, magicInstanceSelector.text(), parametersString) + "."
                } else {
                    return String.format(ability.abilityType.text, magicInstanceSelector.text(), parametersString) + " until end of turn."
                }
            } else {
                return String.format(ability.abilityType.text, parametersString)
            }
        }

    fun getParameter(i: Int): String? {
        if (ability.parameters!!.size > i) {
            return ability.parameters!![i]
        }
        return null
    }

    fun requiresTarget(): Boolean {
        return !ability.targets!!.isEmpty()
    }

    companion object {
        @JvmStatic
        fun getCardInstanceAbilities(card: Card?): List<CardInstanceAbility?> =
            card?.abilities.orEmpty()
                .map { getCardInstanceAbility(it) }

        @JvmStatic
        fun getCardInstanceAbility(ability: Ability?): CardInstanceAbility? =
            ability?.let {
                CardInstanceAbility(it.transpose())
        }
    }
}
